### PR TITLE
Fix: Errant GTID detection on the replicas when they set replication source

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR adds errant GTID detection logic in `setReplicationSourceLocked` call. The intent is that we want to prevent any replica that has an errant GTID from even joining the replication stream. This is because we want to prevent this replica from sending semi-sync ACKs, which can land us into some really annoying and hard to deal with situations. 

To do the errant GTID detection, the tablets first query their own position and then ask the primary for its position. Then they compare the two to see if they have an errant GTID or not. If they do, they fail the RPC without setting the new source.

## Related Issue(s)

- Part of an effort to address the issue https://github.com/vitessio/vitess/issues/16724

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/replication/mysql56_gtid_set.go [hunk 1]
```diff
@@ -467,6 +467,37 @@ func (set Mysql56GTIDSet) SIDBlock() []byte {
...
+	primaryGTIDSet, primaryOk := primaryPosition.GTIDSet.(Mysql56GTIDSet)
+
+	// Currently we only support errant GTID detection for MySQL 56 flavour.
+	if !replicaOk || !primaryOk {
+		return "", nil
+	}
+
+	// Calculate the difference between the replica and primary GTID sets.
+	diffSet := replicaGTIDSet.Difference(primaryGTIDSet)
+	return diffSet.String(), nil
```
Comment (shlomi-noach) [inline go/mysql/replication/mysql56_gtid_set.go:486]: 👍

File: go/mysql/replication/mysql56_gtid_set.go [hunk 2]
```diff
@@ -467,6 +467,37 @@ func (set Mysql56GTIDSet) SIDBlock() []byte {
...
+func (set Mysql56GTIDSet) RemoveUUID(uuid SID) Mysql56GTIDSet {
+	newSet := make(Mysql56GTIDSet)
+	for sid, intervals := range set {
+		if sid == uuid {
+			continue
+		}
+		newSet[sid] = intervals
+	}
+	return newSet
+}
```
Comment (shlomi-noach) [inline go/mysql/replication/mysql56_gtid_set.go:499]: 👍

File: go/mysql/replication/mysql56_gtid_set_test.go [hunk 3]
```diff
@@ -704,3 +704,91 @@ func BenchmarkMySQL56GTIDParsing(b *testing.B) {
...
+		}, {
+			name:               "Primary position parse error",
+			replicaPosition:    "MySQL56/8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-8,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1",
+			primaryPositionStr: "incorrect position",
+			wantErr:            "unknown GTIDSet flavor",
+		}, {
+			name:               "Single errant GTID",
+			replicaPosition:    "MySQL56/8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-8,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1,8bc65cca-3fe4-11ed-bbfb-091034d48bd3:34",
+			primaryPositionStr: "MySQL56/8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-50,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1-30",
+			errantGtidWanted:   "8bc65cca-3fe4-11ed-bbfb-091034d48bd3:34",
```
Comment (shlomi-noach) [inline go/mysql/replication/mysql56_gtid_set_test.go:740]: 👍

File: go/mysql/replication/mysql56_gtid_set_test.go [hunk 4]
```diff
@@ -704,3 +704,91 @@ func BenchmarkMySQL56GTIDParsing(b *testing.B) {
...
+		}, {
+			name:               "Multiple errant GTID",
+			replicaPosition:    "MySQL56/8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-8,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1-32,8bc65cca-3fe4-11ed-bbfb-091034d48bd3:3-35",
+			primaryPositionStr: "MySQL56/8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-50,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1-30,8bc65cca-3fe4-11ed-bbfb-091034d48bd3:34",
+			errantGtidWanted:   "8bc65cca-3fe4-11ed-bbfb-091034d48b3e:31-32,8bc65cca-3fe4-11ed-bbfb-091034d48bd3:3-33:35",
```
Comment (shlomi-noach) [inline go/mysql/replication/mysql56_gtid_set_test.go:745]: 👍

File: go/mysql/replication/mysql56_gtid_set_test.go [hunk 5]
```diff
@@ -704,3 +704,91 @@ func BenchmarkMySQL56GTIDParsing(b *testing.B) {
...
+		name       string
+		initialSet string
+		uuid       string
+		wantSet    string
+	}{
+		{
+			name:       "Remove unknown UUID",
+			initialSet: "8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-8,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1:4-24",
+			uuid:       "8bc65c84-3fe4-11ed-a912-257f0fcde6c9",
+			wantSet:    "8bc65c84-3fe4-11ed-a912-257f0fcdd6c9:1-8,8bc65cca-3fe4-11ed-bbfb-091034d48b3e:1:4-24",
```
Comment (shlomi-noach) [inline go/mysql/replication/mysql56_gtid_set_test.go:775]: 👍

File: go/vt/vttablet/tabletmanager/restore.go [hunk 6]
```diff
@@ -326,7 +323,7 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
 		} else if keyspaceInfo.KeyspaceType == topodatapb.KeyspaceType_NORMAL {
 			// Reconnect to primary only for "NORMAL" keyspaces
 			params.Logger.Infof("Restore: starting replication at position %v", pos)
-			if err := tm.startReplication(context.Background(), pos, originalType); err != nil {
+			if err := tm.startReplication(ctx, pos, originalType); err != nil {
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletmanager/restore.go:326]: 👍

File: go/vt/vttablet/tabletmanager/rpc_replication.go [hunk 7]
```diff
@@ -747,12 +754,32 @@ func (tm *TabletManager) setReplicationSourceLocked(ctx context.Context, parentA
...
 		return vterrors.New(vtrpc.Code_FAILED_PRECONDITION, "Shard primary has empty mysql hostname")
 	}
+	// Errant GTID detection.
+	{
+		// Find the executed GTID set of the tablet that we are reparenting to.
+		// We will then compare our own position against it to verify that we don't
+		// have an errant GTID. If we find any GTID that we have, but the primary doesn't,
+		// we will not enter the replication graph and instead fail replication.
+		var primaryPosition string
+		primaryPosition, err = tm.tmc.PrimaryPosition(ctx, parent.Tablet)
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletmanager/rpc_replication.go:771]: for consistency and simplicity
```suggestion
		primaryPosition, err := tm.tmc.PrimaryPosition(ctx, parent.Tablet)
```

File: go/vt/vttablet/tabletmanager/tm_init.go [hunk 8]
```diff
@@ -1003,21 +1010,47 @@ func (tm *TabletManager) initializeReplication(ctx context.Context, tabletType t
...
+		return "", nil
+	}
+
+	// Find our own executed GTID set and,
+	// the executed GTID set of the tablet that we are reparenting to.
+	// We will then compare our own position against it to verify that we don't
+	// have an errant GTID. If we find any GTID that we have, but the primary doesn't,
+	// we will not enter the replication graph and instead fail replication.
+	var replicaPos replication.Position
+	replicaPos, err = tm.MysqlDaemon.PrimaryPosition(ctx)
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletmanager/tm_init.go:1032]: For consistency and simplicity
```suggestion
	replicaPos, err := tm.MysqlDaemon.PrimaryPosition(ctx)
```

File: go/vt/vttablet/tabletmanager/tm_init.go [hunk 9]
```diff
@@ -1003,21 +1010,47 @@ func (tm *TabletManager) initializeReplication(ctx context.Context, tabletType t
...
+		return "", err
+	}
+
+	primaryPosStr, err = tm.tmc.PrimaryPosition(ctx, currentPrimary.Tablet)
+	if err != nil {
+		return "", err
 	}
+
+	var errantGTIDs string
+	errantGTIDs, err = replication.ErrantGTIDsOnReplica(replicaPos, primaryPosStr)
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletmanager/tm_init.go:1043]:
```suggestion
	errantGTIDs, err := replication.ErrantGTIDsOnReplica(replicaPos, primaryPosStr)
```

**Review by shlomi-noach — APPROVED** [review]
LGTM (with comments). Are we convinced that `setReplicationSourceLocked` is the only path (other than `restore.go` which I've noticed is handled as well)?

File: go/mysql/replication/mysql56_gtid_set.go [hunk 10]
```diff
@@ -467,6 +467,37 @@ func (set Mysql56GTIDSet) SIDBlock() []byte {
 	return buf.Bytes()
 }
 
+// ErrantGTIDsOnReplica gets the errant GTIDs on the replica by comparing against the primary position and UUID.
+func ErrantGTIDsOnReplica(replicaPosition Position, primaryPositionStr string) (string, error) {
```
Comment (deepthi) [inline go/mysql/replication/mysql56_gtid_set.go:471]: It would be more symmetric to pass in the primaryPosition as a `Position` instead of a string.

**Review by deepthi — APPROVED** [review]
We should note in the release notes that for people running with vtop on k8s, this change means that the replica with broken replication will show as unready. There will be some kind of manual intervention needed to clean up these errant replica tablets, even if using the vtorc feature of marking them as drained.

File: go/vt/vttablet/tabletmanager/rpc_replication.go (hunk 7 above)
Comment (GuptaManan100) [reply go/vt/vttablet/tabletmanager/rpc_replication.go]: Fixed

File: go/vt/vttablet/tabletmanager/tm_init.go (hunk 8 above)
Comment (GuptaManan100) [reply go/vt/vttablet/tabletmanager/tm_init.go]: Fixed

File: go/vt/vttablet/tabletmanager/tm_init.go (hunk 9 above)
Comment (GuptaManan100) [reply go/vt/vttablet/tabletmanager/tm_init.go]: Fixed

File: go/mysql/replication/mysql56_gtid_set.go (hunk 10 above)
Comment (GuptaManan100) [reply go/mysql/replication/mysql56_gtid_set.go]: Fixed

**Comment by GuptaManan100** [comment]
Thank you for the reviews! I've addressed all review comments and added the release notes changes as requested. I will merge the PR once the tests are green.

**Pull request merged**